examples/Example2/example2.cu

```diff
@@ -52,26 +204,62 @@ __global__ void DefinePhysicalStepLength(adept::BlockData<track> *block, process
   for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
     // skip particles that are already dead
     track &mytrack = (*block)[i];
+
+    // Experimental limit on number of steps - to avoid 'forever' particles
+    constexpr uint16_t maxNumSteps = 250;  // Configurable in real simulation -- 1000 ?
+    ++  mytrack.num_step;
+    if( mytrack.num_step > maxNumSteps ) {
+       mytrack.status = dead;
+    }
+    
     if (mytrack.status == dead) continue;
 
     // return value (if step limited by physics or geometry) not used for the moment
     // now, I know which process wins, so I add the particle to the appropriate queue
     float physics_step = proclist->GetPhysicsInteractionLength(i, block);
-    float step         = LoopNavigator::ComputeStepAndPropagatedState(mytrack.pos, mytrack.dir, physics_step,
-                                                              mytrack.current_state, mytrack.next_state);
-    mytrack.pos += (step + kPush) * mytrack.dir;
-    if (mytrack.next_state.IsOnBoundary()) {
+    float step= 0.0;
+
+    if( ! BfieldOn ) {
+       step= LoopNavigator::ComputeStepAndPropagatedState(mytrack.pos, mytrack.dir, physics_step,
+                                                          mytrack.current_state, mytrack.next_state);
+       mytrack.pos += (step + kPush) * mytrack.dir;
+       if( step < physics_step ) mytrack.current_process = -1;
+    }    
+    else {
+       step= FieldPropagator__ComputeStepAndPropagatedState(mytrack, physics_step, BzFieldValue
+          );
+       // mytrack.pos = endPosition;   // Already updated in 'mytrack'
+       // mytrack.dir = endDirection;
+       if( step < physics_step ) {
+          // assert( ! mytrack.next_state.IsOnBoundary() && "Field Propagator returned step<phys -- yet boundary!" );
+          mytrack.current_process = -2;
+       }
+    }
+    mytrack.total_length += step;
+
+    if ( mytrack.next_state.IsOnBoundary()) { // ( step == physics_step ) {  // JA-TEMP 
       // For now, just count that we hit something.
       scor->hits++;
+      mytrack.SwapStates();      
     } else {
       // Enqueue track index for later processing.
-      queues[mytrack.current_process]->push_back(i);
+      // assert(mytrack.current_process >= 0);
+      if(mytrack.current_process >= 0)  // JA-TOFIX --- take this out !!! 
```

Comment:
That's a reasonable suggestion.  I will also set a small(er) maximum number of iterations. 
In both cases, though, there will need to be a new state of the particle such as 'transit_in_field" to show that it's integration is ongoing.   

It's like it has reached a 'virtual' boundary in the volume - it will continue in the same volume.  ( The physics processes will not have to be resampled, but if there are it is not ok. )

I will not add these to the current MR - they will come in a future refinement.

Reply:
My comment was a bit misleading: Originally I wondered if this condition is necessary (based on the comment "take this out"). I now understand that, with a maximum number of iterations, this situation can arise and needs to be dealt with. I certainly don't have enough knowledge to recommend one variant over the other - it's not even clear which performance impact this will have, given that control flow will diverge anyway.

For now, it may be best to just go with the easiest solution. From an implementation perspective, that's likely an upper bound on the number of iterations as already coded. I think we can just leave this for the first version.

Reply:
Ensuring that tracking works correctly when field propagation reaches a maximum number of iterations remains an issue.  I will tackle it in a future revision.